### PR TITLE
Optimize EDT

### PR DESCRIPTION
**Pull request by fsx950223**
# Description
@WindQAQ  I need your help since you are good at tensorflow internals.
Original implementation https://github.com/fsx950223/EDT

Brief Description of the PR:
EDT based on http://www.theoryofcomputing.org/articles/v008a019/v008a019.pdf whose time complexity is O(n).
Fixes # (issue)
Fix https://github.com/tensorflow/addons/issues/2399
## Type of change

- [x] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [x] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

If you're adding a bugfix or new feature please describe the tests that you ran to verify your changes:
*

### CODE REVIEW

**Comment by bhack** [comment]
Here the authors's official c++ reference impl http://cs.brown.edu/people/pfelzens/dt/

**Comment by bhack** [comment]
You could add some comment to mention the reference paper/algo and some inline comment like e.g. in https://github.com/giorgiomarcias/distance_transform/blob/master/include/distance_transform/inlines/distance_transform.inl or extracted from the pseudo-code in the paper.

**Comment by fsx950223** [comment]
> You could add a comment to mention the reference paper/algo and some inline comments like e.g. in https://github.com/giorgiomarcias/distance_transform/blob/master/include/distance_transform/inlines/distance_transform.inl or extracted from the pseudo-code in the paper.

I will add it in python docs.

**Comment by bhack** [comment]
> > You could add a comment to mention the reference paper/algo and some inline comments like e.g. in https://github.com/giorgiomarcias/distance_transform/blob/master/include/distance_transform/inlines/distance_transform.inl or extracted from the pseudo-code in the paper.
> 
> I will add it in python docs.

Ok but some inline comments are useful also in the c++ code as you can see in the example and in the pseudocode.

**Comment by fsx950223** [comment]
Good, I will add comments.

**Comment by fsx950223** [comment]
How can I get stack infos by bazel test.
I only get ```Current thread 0x00007f4c9b935700 (most recent call first):
  File "<string>", line 47 in addons_euclidean_distance_transform
  File "/root/.cache/bazel/_bazel_root/e051f2f195071208ea7f081f88730f4f/execroot/__main__/bazel-out/k8-opt/bin/tensorflow_addons/image/image_test.runfiles/__main__/tensorflow_addons/image/distance_transform.py", line 72 in euclidean_dist_transform
  File "/root/.cache/bazel/_bazel_root/e051f2f195071208ea7f081f88730f4f/execroot/__main__/bazel-out/k8-opt/bin/tensorflow_addons/image/image_test.runfiles/__main__/tensorflow_addons/image/tests/distance_transform_test.py", line 65 in test_single_binary_image
  File "/usr/local/lib/python3.6/dist-packages/_pytest/python.py", line 182 in pytest_pyfunc_call
  File "/usr/local/lib/python3.6/dist-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 87 in <lambda>
  File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/local/lib/python3.6/dist-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/local/lib/python3.6/dist-packages/_pytest/python.py", line 1477 in runtest
  File "/usr/local/lib/python3.6/dist-packages/_pytest/runner.py", line 135 in pytest_runtest_call
  File "/usr/local/lib/python3.6/dist-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 87 in <lambda>
  File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/local/lib/python3.6/dist-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/local/lib/python3.6/dist-packages/_pytest/runner.py", line 217 in <lambda>
  File "/usr/local/lib/python3.6/dist-packages/_pytest/runner.py", line 244 in from_call
  File "/usr/local/lib/python3.6/dist-packages/_pytest/runner.py", line 217 in call_runtest_hook
  File "/usr/local/lib/python3.6/dist-packages/_pytest/runner.py", line 186 in call_and_report
  File "/usr/local/lib/python3.6/dist-packages/_pytest/runner.py", line 100 in runtestprotocol
  File "/usr/local/lib/python3.6/dist-packages/_pytest/runner.py", line 85 in pytest_runtest_protocol
  File "/usr/local/lib/python3.6/dist-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 87 in <lambda>
  File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/local/lib/python3.6/dist-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/local/lib/python3.6/dist-packages/_pytest/main.py", line 272 in pytest_runtestloop
  File "/usr/local/lib/python3.6/dist-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 87 in <lambda>
  File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/local/lib/python3.6/dist-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/local/lib/python3.6/dist-packages/_pytest/main.py", line 247 in _main
  File "/usr/local/lib/python3.6/dist-packages/_pytest/main.py", line 191 in wrap_session
  File "/usr/local/lib/python3.6/dist-packages/_pytest/main.py", line 240 in pytest_cmdline_main
  File "/usr/local/lib/python3.6/dist-packages/pluggy/callers.py", line 187 in _multicall
  File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 87 in <lambda>
  File "/usr/local/lib/python3.6/dist-packages/pluggy/manager.py", line 93 in _hookexec
  File "/usr/local/lib/python3.6/dist-packages/pluggy/hooks.py", line 286 in __call__
  File "/usr/local/lib/python3.6/dist-packages/_pytest/config/__init__.py", line 125 in main
  File "/root/.cache/bazel/_bazel_root/e051f2f195071208ea7f081f88730f4f/execroot/__main__/bazel-out/k8-opt/bin/tensorflow_addons/image/image_test.runfiles/__main__/tensorflow_addons/image/tests/run_all_test.py", line 8 in <module>
external/bazel_tools/tools/test/test-setup.sh: line 310: 24547 Aborted                 (core dumped) "${TEST_PATH}" "$@" 2>&1
```
when I failed test cases.

**Comment by fsx950223** [comment]
Now, the implementation only has some bugs in float16 and edge cases.

File: tensorflow_addons/image/tests/distance_transform_test.py [hunk 1]
```diff
@@ -126,5 +126,5 @@ def test_all_zeros():
 def test_all_ones():
     image = tf.ones([10, 10, 1], tf.uint8)
     output = dist_ops.euclidean_dist_transform(image)
-    expected_output = np.full([10, 10, 1], tf.float32.max)
+    expected_output = np.full([10, 10, 1], tf.math.sqrt(tf.float32.max))
```
Comment (bhack) [inline tensorflow_addons/image/tests/distance_transform_test.py:129]: Do you think that we could test the new algo against `scipy.ndimage.distance_transform_edt`?

Comment (bhack) [inline tensorflow_addons/image/tests/distance_transform_test.py:129]: Do you think that we could test the new algo against `scipy.ndimage.distance_transform_edt`?

**Comment by WindQAQ** [comment]
Can you use `@pytest.mark.with_device(["cpu", "gpu"])` on all tests?

**Comment by fsx950223** [comment]
> Can you use `@pytest.mark.with_device(["cpu", "gpu"])` on all tests? Also, I think it now does not take multi channels image into account.

I got ```
cv2.error: OpenCV(4.5.1) /tmp/pip-req-build-_a0ur5ao/opencv/modules/imgproc/src/distransform.cpp:724: error: (-215:Assertion failed) src.type() == CV_8UC1 in function 'distanceTransform'
``` when I run ```python
    src_np = np.round(np.random.rand(512, 512, 3) * 255).astype('uint8')
    cv2.distanceTransform(src_np, cv2.DIST_L2, 0)
```,
So edt doesn't support multi channels?

**Comment by fsx950223** [comment]
I can't pull gcr images in china mainland, could you give me a hand @WindQAQ .

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc [hunk 2]
```diff
@@ -85,11 +97,11 @@ typedef Eigen::GpuDevice GPUDevice;
...
-      const GPUDevice &device, OutputType *output, const InputType &images) \
-      const;                                                                \
+#define DECLARE_FUNCTOR(TYPE)                                            \
+  template <>                                                            \
+  void EuclideanDistanceTransformFunctor<GPUDevice, TYPE>::operator()(   \
+      OpKernelContext *ctx, OutputType *output, const InputType &images) \
+      const;                                                             \
   extern template struct EuclideanDistanceTransformFunctor<GPUDevice, TYPE>
 
 TF_CALL_half(DECLARE_FUNCTOR);
```
Comment (fsx950223) [inline tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc:107]: Why we don't register a uint8 kernel

**Comment by WindQAQ** [comment]
> I can't pull gcr images in china mainland and I don't have a GPU development environment, could you give me a hand @WindQAQ .

We have CI Ubuntu GPU python3. I'm surprised that it can pass the tests on GPU though.

**Comment by fsx950223** [comment]
> > I can't pull gcr images in china mainland and I don't have a GPU development environment, could you give me a hand @WindQAQ .
> 
> We have CI Ubuntu GPU python3. Looks like it can pass all tests.

Yeah, but it's really inefficient.

File: tensorflow_addons/image/tests/distance_transform_test.py (hunk 1 above)
Comment (fsx950223) [reply tensorflow_addons/image/tests/distance_transform_test.py:135]: I keep the same result as opencv's edt.

**Comment by bhack** [comment]
> So edt doesn't support multi channels?

Opencv not but probably `ndimage.distance_transform_edt`

**Comment by WindQAQ** [comment]
@fsx950223 I'm sorry that I'm not aware we only supported channel=1 images previous. Do you think it's reasonable to support multi-channel images?

**Comment by fsx950223** [comment]
> @fsx950223 I'm sorry that I'm not aware we only supported channel=1 images previous. Do you think it's reasonable to support multi-channel images?

It has similar calculation process with batch size, it's more convenient to add multi channels.

**Comment by fsx950223** [comment]
> @fsx950223 I'm sorry that I'm not aware we only supported channel=1 images previous. Do you think it's reasonable to support multi-channel images?

I have implemented multi channels feature, I'm trying to optimize gpu kernel now.

**Comment by fsx950223** [comment]
> @fsx950223 I'm sorry that I'm not aware we only supported channel=1 images previous. Do you think it's reasonable to support multi-channel images?

Do you know what's wrong with my code?
GPU kernel returns:
```E   tensorflow.python.framework.errors_impl.InternalError: Could not synchronize CUDA stream: CUDA_ERROR_LAUNCH_FAILED: unspecified launch failure```

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc [hunk 3]
```diff
@@ -20,15 +20,48 @@ limitations under the License.
...
+
+template <typename T>
+struct EuclideanDistanceTransformFunctor<GPUDevice, T> {
+  typedef typename TTypes<T, 4>::ConstTensor InputType;
+  typedef typename TTypes<T, 4>::Tensor OutputType;
+
+  void operator()(OpKernelContext *ctx, OutputType *output,
+                  const InputType &images) const {
+    auto d = ctx->eigen_device<GPUDevice>();
+    GpuLaunchConfig config = GetGpuLaunchConfig(images.size(), d);
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc:56]: @fsx950223 
Maybe
```python
    GpuLaunchConfig config = GetGpuLaunchConfig(images.dimension(0), d);
```

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc [hunk 4]
```diff
@@ -20,15 +20,48 @@ limitations under the License.
...
 namespace addons {
 
 namespace functor {
 
-// Explicit instantiation of the GPU functor.
-typedef Eigen::GpuDevice GPUDevice;
+template <typename T>
+__global__ void EuclideanDistanceTransformGPUKernel(
+    const int batch_size, const T *input_ptr, const int input_height,
+    const int input_width, const int input_channel, T *output_ptr,
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc:35]: `__restrict__` for input_ptr and output_ptr

File: tensorflow_addons/image/distance_transform.py [hunk 5]
```diff
@@ -32,10 +32,12 @@ def euclidean_dist_transform(
     name: Optional[str] = None,
 ) -> tf.Tensor:
     """Applies euclidean distance transform(s) to the image(s).
+    Based on [Distance Transforms of Sampled Functions]
```
Comment (WindQAQ) [inline tensorflow_addons/image/distance_transform.py:35]: Add an empty line above

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.h [hunk 6]
```diff
@@ -18,78 +18,124 @@ limitations under the License.
...
+namespace generator {
 
 template <typename Device, typename T>
 class EuclideanDistanceTransformGenerator {
  private:
   typename TTypes<T, 4>::ConstTensor input_;
-  int64 height_, width_;
+  int64 height_, width_, channel_, max_;
+
+  void distance(const Eigen::Array<T, Eigen::Dynamic, 1>& f,
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.h:43]: `EIGEN_DEVICE_FUNC`?

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc (hunk 3 above)
Comment (fsx950223) [reply tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc]: Fix the memory issue but crashed.

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc [hunk 7]
```diff
@@ -20,15 +20,49 @@ limitations under the License.
...
 namespace functor {
 
-// Explicit instantiation of the GPU functor.
-typedef Eigen::GpuDevice GPUDevice;
+template <typename T>
+__global__ void EuclideanDistanceTransformGPUKernel(
+    const int batch_size, const T *__restrict__ input_ptr,
+    const int input_height, const int input_width, const int input_channel,
+    T *__restrict__ output_ptr, const int output_height, const int output_width,
+    const int output_channel) {
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc:37]: height width channel are the same for input and output.

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc [hunk 8]
```diff
@@ -20,15 +20,49 @@ limitations under the License.
...
+  void operator()(OpKernelContext *ctx, OutputType *output,
+                  const InputType &images) const {
+    auto d = ctx->eigen_device<GPUDevice>();
+    GpuLaunchConfig config = GetGpuLaunchConfig(output->size(), d);
+    TF_CHECK_OK(GpuLaunchKernel(
+        EuclideanDistanceTransformGPUKernel<T>, config.block_count,
+        config.thread_per_block, 0, d.stream(), int(images.dimension(0)),
+        images.data(), int(images.dimension(1)), int(images.dimension(2)),
+        int(images.dimension(3)), output->data(), int(output->dimension(1)),
+        int(output->dimension(2)), int(output->dimension(3))));
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc:63]: can we use `static_cast<int>` instead?

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc (hunk 3 above)
Comment (WindQAQ) [reply tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc]: Do you have other logs?

Comment (WindQAQ) [reply tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc]: Maybe run on colab with GDB.

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc (hunk 8 above)
Comment (fsx950223) [reply tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc:63]: Should all type cast to be replaced with static_cast?

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc (hunk 2 above)
Comment (WindQAQ) [reply tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc:110]: Actually, I don't know, but we can do it now :-)

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc (hunk 3 above)
Comment (fsx950223) [reply tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc]: Here https://github.com/tensorflow/tensorflow/blob/dec8e0b11f4f87693b67e125e67dfbc68d26c205/tensorflow/core/kernels/dilation_ops_gpu.cu.cc#L196

**Comment by fsx950223** [comment]
I would disable gpu kernels until find the solution.

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc (hunk 3 above)
Comment (WindQAQ) [reply tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op_gpu.cu.cc]: Hmm, it's weird.

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.h [hunk 9]
```diff
@@ -18,78 +18,126 @@ limitations under the License.
...
-  EuclideanDistanceTransformGenerator(typename TTypes<T, 4>::ConstTensor input)
+  EIGEN_DEVICE_FUNC EIGEN_ALWAYS_INLINE EuclideanDistanceTransformGenerator(
+      const typename TTypes<T, 4>::ConstTensor& input)
       : input_(input) {
-    height_ = input_.dimension(1);
-    width_ = input_.dimension(2);
+    height_ = input.dimension(1);
+    width_ = input.dimension(2);
+    channel_ = input.dimension(3);
+    max_ = std::max(height_, width_);
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.h:88]: Can you use max instead? std::max is not a device function.

**Comment by fsx950223** [comment]
I guess eigen can't work with cuda directly, I should use pointer instead.

**Comment by fsx950223** [comment]
Gpu float16 and float64 GPU kernels still don't work, is it necessary to add float16 and float64 kernels?

**Comment by fsx950223** [comment]
> > Gpu float16 and float64 GPU kernels still don't work, is it necessary to add float16 and float64 kernels?
> 
> Does float32 work? (Sorry that I accidentally edit your comment)

Yes

**Comment by WindQAQ** [comment]
It's ok for me. Can you keep them the same on CPU but do not register/specialize them on GPU? Probably also leave some comments.

**Comment by fsx950223** [comment]
Done

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.h [hunk 10]
```diff
@@ -18,78 +18,120 @@ limitations under the License.
...
- public:
-  EIGEN_DEVICE_FUNC EIGEN_ALWAYS_INLINE
-  EuclideanDistanceTransformGenerator(typename TTypes<T, 4>::ConstTensor input)
-      : input_(input) {
-    height_ = input_.dimension(1);
-    width_ = input_.dimension(2);
+typedef Eigen::ThreadPoolDevice CPUDevice;
+typedef Eigen::GpuDevice GPUDevice;
+
+#define GETINDEX(i, j, k, c) \
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.h:35]: `GET_INDEX`?

File: tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc [hunk 11]
```diff
@@ -104,9 +121,9 @@ TF_CALL_double(DECLARE_FUNCTOR);
                               .TypeConstraint<TYPE>("dtype"),       \
                           EuclideanDistanceTransform<GPUDevice, TYPE>)
 
-TF_CALL_half(REGISTER);
+// TF_CALL_half(REGISTER);  // float16 kernel crashs ont GPU
 TF_CALL_float(REGISTER);
-TF_CALL_double(REGISTER);
+// TF_CALL_double(REGISTER);  // float64 kernel crashs ont GPU
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/euclidean_distance_transform_op.cc:126]: NIT: ont GPU -> on GPU

**Review by WindQAQ — COMMENTED** [review]
LGTM. I see you are using int32 indexing. Can you either switch to int64 or throw an error when number of elements of input is larger than int32 max?

**Comment by WindQAQ** [comment]
Hi @fsx950223, changing `index` is not enough. We have to change `batch`, `height`, `width`, `channel`, and all other variables associated to them to int64.

**Comment by fsx950223** [comment]
> height

It's impossible. If so, Height*width may out of range of int64.

**Comment by WindQAQ** [comment]
> > height
> 
> It's impossible. If so, height*width may out of range of int64.

I think there is some extreme case like (1, int32_max + 1, 1, 3).

**Comment by fsx950223** [comment]
> > > height
> > 
> > 
> > It's impossible. If so, height*width may out of range of int64.
> 
> I think there is some extreme case like `(b, h, w, c) = (1, int32_max + 1, 1, 1)`.

Unfortunately, it's crashed on GPU.

**Comment by WindQAQ** [comment]
Cool. Can we all use 32 bits indexing including the value obtained from GET_INDEX, and then throw an error when input number of elements is larger then int32 max in OpKernel?

**Comment by fsx950223** [comment]
> Cool. Can we all use 32 bits indexing including the value obtained from GET_INDEX, and then throw an error when input number of elements is larger then int32 max in OpKernel?

The max value of int32 is 2^31-1?
I found tensorflow/core/kernels/image/resize_bilinear_op_gpu.cu.cc use 2^32-1 as int32 max value.

**Comment by WindQAQ** [comment]
> > Cool. Can we all use 32 bits indexing including the value obtained from GET_INDEX, and then throw an error when input number of elements is larger then int32 max in OpKernel?
> 
> The max value of int32 is 2^31-1?
> I found tensorflow/core/kernels/image/resize_bilinear_op_gpu.cu.cc use 2^32-1 as int32 max value.

Yes it is.

**Review by WindQAQ — APPROVED** [review]

**Comment by WindQAQ** [comment]
@fsx950223 Thanks for the optimization 😃 It would be nice if you can figure out why float16 and float64 will fail on GPU.
